transaction/DirectedGraph.hpp

```diff
@@ -0,0 +1,173 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_TRANSACTION_DIRECTED_GRAPH_HPP_
+#define QUICKSTEP_TRANSACTION_DIRECTED_GRAPH_HPP_
+
+#include <algorithm>
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <stack>
+#include <vector>
+
+#include "transaction/Transaction.hpp"
+#include "utility/Macros.hpp"
+
+namespace quickstep {
+
+namespace transaction {
+
+/** \addtogroup Transaction
+ *  @{
+ */
+
+/**
+ * @brief Class for representing a directed graph.
+ *        Vertices are transaction ids, edges are
+ *        wait-for relations.
+ */
+class DirectedGraph {
+ public:
+  using NodeId = std::uint64_t;
+
+  /**
+   * @brief Default constructor
+   */
+  DirectedGraph() {}
+
+  /**
+   * @brief Adds node with given transaction id.
+   * @warning It does not check whether transaction id
+   *          is already in the graph.
+   * @warning Pointer ownership will pass to the graph,
+   *          threfore it should not be deleted.
+   *
+   * @param data Pointer to the transaction id that
+   *        will be contained in the node.
+   *
+   * @return Id of the newly created node.
+   */
+  inline NodeId addNode(TransactionId *data) {
+    nodes_.emplace_back(data);
+    return nodes_.size() - 1;
+  }
+
+  /**
+   * @brief Adds an edge between nodes.
+   * @warning Does not check arguments are legit.
```

Comment:
Similar to the comments in ``addNode`` method, you can implement two methods -- an unchecked version and a checked version.